### ISSUE TITLE
old joinpath issue was symptom of compiler bug

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#33622]
As noted in https://github.com/JuliaLang/julia/issues/33455#issuecomment-540113905, the old `joinpath` bug on Windows may have been a symptom of a compiler bug. The symptom is now fixed by @musm's rewrite of `joinpath` to be more efficient and no longer use recursion, but the compiler bug was not fixed so I'm opening this issue to track that. The reproducer is to checkout a version of Julia before https://github.com/JuliaLang/julia/pull/33477 (or revert it).

**Comment by vtjnash** [comment]
I can confirm this still seems broken, with:
```
julia> joinpath("\\\\server\\share","b")
"\\\\server\\shareb"
```

``````diff
commit 609dd7de77e1cd3b78a8a3680aa8dae871b8bb8e
Author: Jameson Nash <jameson@juliacomputing.com>
Date:   Wed Mar 17 16:07:26 2021 -0400

    Revert "Fix joinpath on Windows and improve joinpath on linux (#33477)"
    
    This reverts commit e14616a6887d66f58ede0b3efb13b5aa025d8dfd.

diff --git a/base/path.jl b/base/path.jl
index 1a7c2dc550..8530665dd4 100644
--- a/base/path.jl
+++ b/base/path.jl
@@ -208,8 +208,13 @@ function splitext(path::String)
     a*m.captures[1], String(m.captures[2])
 end
 
-# NOTE: deprecated in 1.4
-pathsep() = path_separator
+function pathsep(paths::AbstractString...)
+    for path in paths
+        m = match(path_separator_re, String(path))
+        m !== nothing && return m.match[1:1]
+    end
+    return path_separator
+end
 
 """
     splitpath(path::AbstractString) -> Vector{String}
@@ -251,85 +256,33 @@ function splitpath(p::String)
     return out
 end
 
-joinpath(path::AbstractString)::String = path
-
-if Sys.iswindows()
-
-function joinpath(path::AbstractString, paths::AbstractString...)::String
-    result_drive, result_path = splitdrive(path)
-
-    local p_drive, p_path
-    for p in paths
-        p_drive, p_path = splitdrive(p)
-
-        if startswith(p_path, ('\\', '/'))
-            # second path is absolute
-            if !isempty(p_drive) || !isempty(result_drive)
-                result_drive = p_drive
-            end
-            result_path = p_path
-            continue
-        elseif !isempty(p_drive) && p_drive != result_drive
-            if lowercase(p_drive) != lowercase(result_drive)
-                # different drives, ignore the first path entirely
-                result_drive = p_drive
-                result_path = p_path
-                continue
-            end
-        end
-
-        # second path is relative to the first
-        if !isempty(result_path) && result_path[end] ∉ ('\\', '/')
-            result_path *= "\\"
-        end
-
-        result_path = result_path * p_path
-    end
-
-    # add separator between UNC and non-absolute path
-    if !isempty(p_path) && result_path[1] ∉ ('\\', '/') && !isempty(result_drive) && result_drive[end] != ':'
-        return result_drive * "\\" * result_path
-    end
-
-    return result_drive * result_path
-end
-
-else
-
-function joinpath(path::AbstractString, paths::AbstractString...)::String
-    for p in paths
-        if isabspath(p)
-            path = p
-        elseif isempty(path) || path[end] == '/'
-            path *= p
-        else
-            path *= "/" * p
-        end
-    end
-    return path
-end
-
-end # os-test
-
 """
-    joinpath(parts::AbstractString...) -> String
+    joinpath(parts...) -> AbstractString
 
 Join path components into a full path. If some argument is an absolute path or
 (on Windows) has a drive specification that doesn't match the drive computed for
 the join of the preceding paths, then prior components are dropped.
 
-Note on Windows since there is a current directory for each drive, `joinpath("c:", "foo")`
-represents a path relative to the current directory on drive "c:" so this is equal to "c:foo",
-not "c:\\foo". Furthermore, `joinpath` treats this as a non-absolute path and ignores the drive
-letter casing, hence `joinpath("C:\\A","c:b") = "C:\\A\\b"`.
-
 # Examples
 ```jldoctest
 julia> joinpath("/home/myuser", "example.jl")
 "/home/myuser/example.jl"
 ```
 """
-joinpath
+joinpath(a::AbstractString, b::AbstractString, c::AbstractString...) = joinpath(joinpath(a,b), c...)
+joinpath(a::AbstractString) = a
+
+function joinpath(a::String, b::String)
+    isabspath(b) && return b
+    A, a = splitdrive(a)
+    B, b = splitdrive(b)
+    !isempty(B) && A != B && return string(B,b)
+    C = isempty(B) ? A : B
+    isempty(a)                              ? string(C,b) :
+    occursin(path_separator_re, a[end:end]) ? string(C,a,b) :
+                                              string(C,a,pathsep(a,b),b)
+end
+joinpath(a::AbstractString, b::AbstractString) = joinpath(String(a), String(b))
 
 """
     normpath(path::AbstractString) -> String
``````

**Comment by vtjnash** [comment]
The comment in #33455 was simply wrong: it was a bug in the old algorithm–which was not the code in that comment–and ignore the claim to the contrary. This is fixed by #33477.
